src/elf/dyn.rs

```diff
@@ -497,12 +502,9 @@ macro_rules! elf_dyn_std_impl {
                     if phdr.p_type == PT_DYNAMIC {
                         let filesz = phdr.p_filesz as usize;
                         let dync = filesz / SIZEOF_DYN;
-                        let mut bytes = vec![0u8; filesz];
+                        let mut dyns = vec![Dyn::default(); dync];
                         try!(fd.seek(Start(phdr.p_offset as u64)));
-                        try!(fd.read(&mut bytes));
-                        let bytes = unsafe { slice::from_raw_parts(bytes.as_ptr() as *mut Dyn, dync) };
-                        let mut dyns = Vec::with_capacity(dync);
-                        dyns.extend_from_slice(bytes);
+                        try!(fd.read(dyns.as_mut_bytes()));
                         dyns.dedup();
```

Comment:
So for the dynamic array, the correct way to read them is [to terminate](https://github.com/m4b/goblin/blob/8188a8a15017cbdadfdbb807e08db968f1c634c9/src/elf/dyn.rs#L379) on the first `DT_NULL` tag, which is just `0x0`, which also fills up the remainder of the dynamic array (e.g., `phdr.p_filesz`) sometimes if there is padding, so occasionally there are several `DT_NULL` entries; the `dedup` was me being lazy here, but I don't think this endpoint has been used by any downstream consumer in very likely a long time, so it doesn't really matter. I should probably just remove it next release, as it isn't a very useful function.